codegen/src/types/substitutes.rs

```diff
@@ -0,0 +1,237 @@
+// Copyright 2019-2023 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+use crate::CratePath;
+use proc_macro_error::abort;
+use std::{
+    borrow::Cow,
+    collections::HashMap,
+};
+use syn::{
+    parse_quote,
+    spanned::Spanned as _,
+};
+
+use super::TypePath;
+
+#[derive(Debug)]
+pub struct TypeSubstitutes {
+    inner: HashMap<PathSegments, syn::Path>,
+    params: HashMap<PathSegments, TypeParamMapping>,
```

Comment:
So I think at the moment, for some type substitute, we can look up the param mapping in the `params` entry. So either `params.get(path) == None` or `params.get(path) == Some(TypeParamMapping::None)` I guess mean the same?

I wonder whether it makes sense to merge these so that we have more like:

```rust
inner: HashMap<PathSegments, (syn::Path, TypeParamMapping)>
```

And then for any path, we always have a type param mapping which is either `TypeParamMapping::None` or `Specified`. I geuss all of the "default" type substitutes would then have `None` to mean "copy over any types as is"

Reply:
Yeah, this makes sense; I forgot to merge those two in the final PR form. Thanks for applying the fix yourself!